[#3]; feat: 게시물 생성, 수정, 삭제 기능 추가

Sennyru opened the pull request:
Closes #3

# 구현 내용
API 명세는 README.md를 참고해 주세요.

## 게시물 생성
`/api/posts`로 POST 요청을 보내면 DB에 게시물이 생성됩니다.
사용자 ID가 존재하지 않거나, PW가 맞지 않는 경우 400 Bad Request를 보냅니다.
성공하면 201 Created를 보냅니다.

## 게시물 수정
`/api/posts/{게시물 ID}`로 PUT 요청을 보내면 DB에 있는 게시물의 제목과 내용이 수정됩니다.
게시물 ID가 존재하지 않거나, 사용자 ID가 존재하지 않거나, PW가 맞지 않거나, 게시물 작성자가 그 사용자가 아닐 경우 400 Bad Request를 보냅니다.
성공하면 200 Ok를 보냅니다.

## 게시물 삭제
`/api/posts/{게시물 ID`로 DELETE 요청을 보내면 DB에 있는 게시물이 삭제됩니다.
게시물 ID가 존재하지 않거나, 사용자 ID가 존재하지 않거나, PW가 맞지 않거나, 게시물 작성자가 그 사용자가 아닐 경우 400 Bad Request를 보냅니다.
성공하면 204 No Content를 보냅니다.

# 느낀 점
회원가입(#2) 만들 때 기본적인 틀을 다 만들어 놓은 상태라서 추가 기능을 구현하는 데 별로 어려움이 없었다.

# 새로 알게 된 것들
- 외래 키 지정
- Path Variable
- POST 이외의 다른 HTTP 요청들

Sennyru committed:
[#3]; feat: 게시물 생성 API 구현

- `/api/posts` 경로로 사용자 ID, PW, 게시물 제목과 내용을 전달하면 게시물을 생성할 수 있음
- 요청 포맷은 README.md의 B번 참고
- 존재하지 않는 계정(ID)이거나 PW가 일치하지 않는 경우 500 Bad Request 응답

Sennyru committed:
[#3]; feat: 게시물 수정 API 구현

- `/api/posts/{변경할 게시물 ID}` 경로로 사용자 ID, PW, 게시물 제목과 내용을 전달하면 게시물 내용을 수정할 수 있음
- 요청 포맷은 README.md의 C번 참고
- 존재하지 않는 계정(ID)이거나, PW가 일치하지 않는 경우 500 Bad Request 응답
- 해당 게시물 ID가 없거나, 게시물 작성자와 다른 사용자가 수정을 요청해도 500 응답

Sennyru committed:
[#3]; feat: 게시물 삭제 API 구현

- `/api/posts/{삭제 게시물 ID}` 경로로 사용자 ID, PW를 전달하면 게시물을 삭제할 수 있음
- 요청 포맷은 README.md의 D번 참고
- 존재하지 않는 계정(ID)이거나, PW가 일치하지 않거나, 해당 게시물 ID가 없거나, 게시물 작성자와 다른 사용자가 삭제를 요청하면 500 Bad Request 응답
- 삭제 완료 시 204 No Content 응답
- 추후 댓글 기능을 추가했을 때, 게시물 삭제 시 댓글도 삭제되게 해야 함

seulnan commented on src/main/java/com/sennyru/onboarding/handler/GlobalExceptionHandler.java:
위에서는 getMessage를 쓰고 여기서는 메시지를 하드코딩한 이유가 있나요?

Sennyru replied:
확인 결과, `DataIntegrityViolationException`은 스프링에서 자체적으로 만드는 예외라서 메시지 형식이 다르고 심지어 영어입니다. 그래서 getMessege를 쓰지 않았고, 대신 한국어로 메시지를 새로 작성해서 넣었습니다.

seulnan replied:
그러면 도메인의미를 담은 예외클래스를 만들면 getmessage로 통일성있는 코드도 짜고 기존의 작동하던 방식도 안해치는거아닌가요?

Sennyru replied:
문제를 어떻게 해결할까 찾아보니까 예외 전환이라는 기법을 알아냈습니다.
```java
try {
    Member savedMember = memberRepository.save(member);
} catch (DataIntegrityViolationException e) {
    throw new EmailAlreadyExistsException("이미 사용 중인 이메일입니다.");
}
```
이런 식으로 Service 측에서 미리 DataIntegrityViolationException을 잡고, 실제로는 다른 예외를 던지게 하면 해결할 수 있을 것 같습니다. 어떤가요?
도메인 의미를 담은 예외 클래스로 해결한다는 것이 이런 느낌인가요?

seulnan replied:
음 그냥 저는 바로 throw EmailAlreadyExistsException 하는걸 생각했는데 준혁님 코드가 예외클래스간 부모구조가 잘 잡혀있어서 더 좋은것같아요. 다만, 무분별한 try-catch는 지양해야한다는 점만 조심해주세요!(현재는 괜찮음)

seulnan reviewed (commented):
고생하셨습니다 궁금한게 bad request는 500이 아니라 400아닌가요?

Sennyru commented:
> 고생하셨습니다 궁금한게 bad request는 500이 아니라 400아닌가요?

헉!! 그러네요. 수정하겠습니다.
지금 보니까 최근 커밋들에도 500이라고 잘못 써놨네요. 코멘트를 달아놓겠습니다.

Sennyru committed:
[#3]; refactor: 게시물·사용자 검증 관련 중복 코드 제거

게시물 관리 로직에서 사용되는 중복된 인증 절차를 따로 분리하여 코드를 절약함.
세부적인 절차는 아래와 같고, 괄호로 묶인 공통 과정을 각각 메서드로 분리함.
- 생성: (존재하는 이메일인가 -> 비밀번호가 일치하는가)
- 수정: (존재하는 이메일인가 -> 비밀번호가 일치하는가) -> (게시물이 존재하는가 -> 게시물 작성자가 맞는가)
- 삭제: (존재하는 이메일인가 -> 비밀번호가 일치하는가) -> (게시물이 존재하는가 -> 게시물 작성자가 맞는가)

Sennyru commented on src/main/java/com/sennyru/onboarding/service/PostService.java:
```diff
@@ -0,0 +1,91 @@
+
+        if (!passwordEncoder.matches(requestDto.password(), member.getPassword())) {
+            throw new IllegalArgumentException("비밀번호가 일치하지 않습니다.");
+        }
```
메서드를 분리했습니다. ([해당 커밋](https://github.com/Sennyru/Backend-Study-Onboarding/pull/4/commits/6c05fc77d9805c49cac5f25a06d11c225dfc7e99))

Sennyru commented on src/main/java/com/sennyru/onboarding/service/PostService.java:
```diff
@@ -0,0 +1,91 @@
+
+        if (!Objects.equals(post.getMember().getId(), member.getId())) {
+            throw new IllegalStateException("게시물 수정 권한이 없습니다.");
+        }
```
메서드를 분리했습니다. ([해당 커밋](https://github.com/Sennyru/Backend-Study-Onboarding/pull/4/commits/6c05fc77d9805c49cac5f25a06d11c225dfc7e99))

Sennyru committed:
[#3]; refactor: 도메인 예외 처리 도입

- `IllegalStateException`, `IllegalArgumentException` 등 그 자체로는 구체적인 원인을 알기 힘든 예외 대신 문제 상황에 알맞는 예외를 사용하여 코드 가독성 개선
- 서비스에서 발생하는 모든 예외는 DomainException을 상속받게 함
- DomainException에 HTTP 상태 코드 필드를 추가함으로써 예외 핸들러의 코드를 획기적으로 줄임
- 회원가입(`MemberService.signup()`) 시 이미 가입된 이메일인지 체크하는 부분을 DataIntegrityViolationException을 try-catch로 잡아내는 방식으로 변경
  - 스프링에서 발생하는 예외를 DomainException을 상속한 예외로 전환
  - 이를 통해 레이스 컨디션이 해결되는지 추후 확인해볼 수 있을 듯
